### PR TITLE
Changed OSX Docker workflow to expect the webpack server to be run on the host machine

### PR DESCRIPTION
**Pull request by gsidebo**
#### What are the relevant tickets?

#307 

#### What's this PR do?

Changes our workflow so OSX users will run the webpack dev server on their host machines rather than inside the Docker container. This eliminates the need for docker-osx-dev.

#### Where should the reviewer start?

Start at `docker-compose.yml` for the config changes, `micromasters/utils.py` for the code changes

#### How should this be manually tested?

OSX users should run `./webpack_dev_server.sh --install` in one terminal tab, and `docker-compose up`. Linux users can just run the latter command. We'll definitely want a Linux user to test this since I hacked on this in OSX. 

#### Any background context you want to provide?

**Major implications:**

1. (as mentioned) OSX users will need to run the webpack server locally in a separate terminal tab, and can do so by running the helper script `./webpack_dev_server.sh`
1. (as mentioned) docker-osx-dev is no longer needed
1. OSX users will now need to add the following to the `.env` file: `WEBPACK_DEV_SERVER_HOST='localhost'`. Before now, webpack was assumed to be running in the container with the web server.
1. The 'watch' service in `docker-compose.yml` now invokes the helper script to start the server. In the case of OSX users, 'watch' will do nothing. I couldn't figure out a good way to entirely exclude the 'watch' service for OSX users that wouldn't require some annoying flags to be specified every single time the `docker-compose` command was used. Any feedback on that is quite welcome.

### CODE REVIEW

**Comment by gsidebo** [comment]
forgot README updates. working on that now...

**Comment by gsidebo** [comment]
some things i'm not sure about:

- the new file `micromasters/utils.py` containing the functions that provide the webpack host/url. couldn't think of a better place for them
- those functions requiring `request` to be passed in. kind of a code smell. is there a cleaner way to determine the machine's IP in django than grabbing it from the request?
- like i said in the first comment, the 'watch' service is still available when running from OSX and it just does nothing
- the webpack_dev_server script is a bit awkward in general

**Comment by giocalitri** [comment]
There are some linting errors.
Please fix

File: micromasters/settings.py [hunk 1]
```diff
@@ -257,7 +257,8 @@ def get_var(name, default):
 
 # Request files from the webpack dev server
 USE_WEBPACK_DEV_SERVER = get_var('MICROMASTERS_USE_WEBPACK_DEV_SERVER', False)
-WEBPACK_SERVER_URL = get_var('MICROMASTERS_WEBPACK_SERVER_URL', 'http://{host}:8078')
+WEBPACK_DEV_SERVER_HOST = get_var('WEBPACK_DEV_SERVER_HOST', '')
```
Comment (giocalitri) [inline micromasters/settings.py]: why an empty default for this?

File: webpack_dev_server.sh [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#       the webpack server on the host machine rather than the container.
+
+if [[ -z "$WEBPACK_DEV_SERVER_HOST" || "$IS_OSX_HOST_MACHINE" == 'true' ]] ; then
+    if [[ "$1" == "--install" ]] ; then
+      npm install --no-bin-links && echo "Finished npm install"
+    fi
+    # Start the webpack dev server on the appropriate host and port
+    node ./node_modules/webpack-dev-server/bin/webpack-dev-server.js --config webpack.config.dev.js -d \
+        --content-base ./static --host "$WEBPACK_HOST" --port "$WEBPACK_PORT" --progress --inline --hot
+fi
```
Comment (giocalitri) [inline webpack_dev_server.sh]: empty line at the end of the file

**Comment by giocalitri** [comment]
I think I need to talk to you in person tomorrow morning to fully understand this PR

**Comment by giocalitri** [comment]
In the meantime, please rebase and fix conflicts

File: micromasters/settings.py (hunk 1 above)
Comment (gsidebo) [inline micromasters/settings.py]: when this isn't specified, we want the webpack server host to be the same as the django host. before this PR, we were pulling the django host value out of the request in our views (`request.get_host()`). i kept that approach. when `WEBPACK_DEV_SERVER_HOST` is blank, we get the host via `request.get_host()`. i mentioned above that if there's an easy way to get the django host that doesn't involve pulling it out of the request, we should go with that

File: ui/views_test.py [hunk 3]
```diff
@@ -91,13 +91,14 @@ def test_authenticated_user_doesnt_redirect(self):
     def test_webpack_url(self):
```
Comment (giocalitri) [inline ui/views_test.py]: I know you did not write this test, but (as note for the future), it is better to use `ddt` or `@pytest.mark.parametrize` for multiple inputs for a test.

File: webpack_dev_server.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+set -e
+
+if [[ $(uname -s) == "Darwin" ]] ; then
+  IS_OSX_HOST_MACHINE="true"
+else
+  IS_OSX_HOST_MACHINE="false"
```
Comment (giocalitri) [inline webpack_dev_server.sh]: does the double quote vs single quote make a difference in bash?

**Comment by giocalitri** [comment]
I have an additional question: on Mac, where the npm packages are installed? in the current directory or in a global/system one?

**Comment by giocalitri** [comment]
First pass done: overall it looks good: I have only a couple of questions and few minor comments

File: webpack_dev_server.sh (hunk 4 above)
Comment (noisecapella) [inline webpack_dev_server.sh]: Only if there are variables inside the string

File: webpack_dev_server.sh [hunk 5]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+set -e
```
Comment (noisecapella) [inline webpack_dev_server.sh]: Can you use `set -euf -o pipefail` here? See https://sipb.mit.edu/doc/safe-shell/

**Comment by gsidebo** [comment]
@giocalitri 

> I have an additional question: on Mac, where the npm packages are installed? in the current directory or in a global/system one?

as @aliceriot kindly reminded me, npm packages are installed locally (ie: in the same directory as the package.json) by default. you need to set the `-g` flag explicitly to install them on the entire system

> In the meantime, please rebase ...

Just want to make sure: when you say 'rebase', you mean that I should squash all the branch commits to one commit, right?

**Comment by giocalitri** [comment]
No, when I say `rebase`, there are conflicts between this branch and master, so you should rebase to fix them.

**Comment by gsidebo** [comment]
@giocalitri just rebased onto master

**Comment by giocalitri** [comment]
I get this error
```
gioMacMit:micromasters gdimilia$ ./webpack_dev_server.sh
./webpack_dev_server.sh: line 18: WEBPACK_DEV_SERVER_PORT: unbound variable
```

**Comment by gsidebo** [comment]
@giocalitri fixed

**Comment by aliceriot** [comment]
not sure if it's ready for linux testing yet, but I just checked out this branch and did `docker-compose up` with no problems.

**Comment by gsidebo** [comment]
@aliceriot it's ready, and thank you! you're way ahead of me. i'll merge this when i get the +1

**Comment by giocalitri** [comment]
from the OS X point of view this is good to go. 👍 
we might need a second opinion from a linux user ( @aliceriot ?)

**Comment by giocalitri** [comment]
please squash the commits

**Comment by gsidebo** [comment]
yeah @aliceriot some very minor changes were made since you tested. should be good to go, but it certainly wouldn't hurt if you have it one more try.

**Comment by aliceriot** [comment]
looks good to me! everything's working just fine.

**Pull request merged**